python-lib/custom_gluon_models/arima.py

```diff
@@ -0,0 +1,134 @@
+from gluonts.model.estimator import Estimator
+from gluonts.model.forecast import SampleForecast
+from gluonts.model.predictor import RepresentablePredictor
+from gluonts.support.pandas import frequency_add
+from gluonts.core.component import validated
+from gluonts.time_feature import get_seasonality
+from custom_gluon_models.utils import cast_kwargs
+from constants import TIMESERIES_KEYS
+import pmdarima as pm
+import numpy as np
+from safe_logger import SafeLogger
+from tqdm import tqdm
+
+
+logger = SafeLogger("Forecast plugin - ARIMA")
+
+
+class ArimaPredictor(RepresentablePredictor):
+    """
+    An abstract predictor that can be subclassed by models that are not based
+    on Gluon. Subclasses should have @validated() constructors.
+    (De)serialization and value equality are all implemented on top of the
+    @validated() logic.
+
+    Parameters
+    ----------
+    prediction_length
+        Prediction horizon.
+    freq
+        Frequency of the predicted data.
+    """
+
+    # TODO implement custom serializer
+
+    @validated()
+    def __init__(self, prediction_length, freq, trained_models, lead_time=0):
+        super().__init__(freq=freq, lead_time=lead_time, prediction_length=prediction_length)
+        self.trained_models = trained_models
+
+    def predict(self, dataset, **kwargs):
+        """
+
+        Args:
+            dataset (gluonts.dataset.common.Dataset): Dataset after wich to predict forecasts.
+
+        Yields:
+            SampleForecast of predictions.
+        """
+        logger.info("Prediction timeseries ...")
+        for i, item in tqdm(enumerate(dataset)):
+            yield self.predict_item(item, self.trained_models[i])
+
+    def predict_item(self, item, trained_model):
+        """Compute quantiles using the confidence intervals of auto_arima.
+
+        Args:
+            item (DataEntry): One timeseries.
+            trained_model (list): List of trained auto_arima models.
+
+        Returns:
+            SampleForecast of quantiles.
+        """
+        start_date = frequency_add(item["start"], len(item["target"]))
+
+        prediction_external_features = self._set_prediction_external_features(item)
+
+        samples = []
+        for alpha in np.arange(0.02, 1.01, 0.02):
+            confidence_intervals = trained_model.predict(n_periods=self.prediction_length, X=prediction_external_features, return_conf_int=True, alpha=alpha)[1]
+            samples += [confidence_intervals[:, 0], confidence_intervals[:, 1]]
+
+        return SampleForecast(samples=np.stack(samples), start_date=start_date, freq=self.freq)
+
+    def _set_prediction_external_features(self, item):
+        prediction_external_features = None
+        if TIMESERIES_KEYS.FEAT_DYNAMIC_REAL_COLUMNS_NAMES in item:
+            prediction_external_features = item[TIMESERIES_KEYS.FEAT_DYNAMIC_REAL][:, -self.prediction_length :].T
+        return prediction_external_features
+
+
+class ArimaEstimator(Estimator):
```

Comment:
To be exact, that model shouldn't be named Arima, but AutoARIMA. The reason is that Arima has a specific meaning in the literature. If you look at https://alkaline-ml.com/pmdarima/modules/classes.html there is one class for ARIMA and another for AutoARIMA.